allennlp/models/encoder_decoders/composed_seq2seq.py

```diff
@@ -0,0 +1,130 @@
+        Embedders for source side sequences
+    encoder : ``Seq2SeqEncoder``, required
+        The encoder of the "encoder/decoder" model
+    decoder : ``SeqDecoder``, required
+        The decoder of the "encoder/decoder" model
+    tied_source_embedder_key : ``str``, optional (default=``None``)
+        If specified, this key is used to obtain token_embedder in `source_text_embedder` and
+        the weights are shared/tied with the decoder's target embedding weights.
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+
+    def __init__(self,
+                 vocab: Vocabulary,
+                 source_text_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 decoder: SeqDecoder,
+                 tied_source_embedder_key: Optional[str] = None,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None)-> None:
+
+        super(ComposedSeq2Seq, self).__init__(vocab, regularizer)
+
+        self._source_text_embedder = source_text_embedder
+        self._encoder = encoder
+        self._decoder = decoder
+
+        if self._encoder.get_output_dim() != self._decoder.get_output_dim():
+            raise ConfigurationError(f"Encoder output dimension {self._encoder.get_output_dim()} should be"
+                                     f" equal to decoder dimension {self._decoder.get_output_dim()}.")
+        if tied_source_embedder_key:
+            # Very ugly way to tie Embeddings
+            if not isinstance(self._source_text_embedder, BasicTextFieldEmbedder):
+                raise ConfigurationError("Unable to tie embeddings,"
+                                         "Source text embedder is not an instance of `BasicTextFieldEmbedder`.")
+            # pylint: disable=protected-access
+            source_embedder = self._source_text_embedder._token_embedders[tied_source_embedder_key]
+            if not isinstance(source_embedder, Embedding):
+                raise ConfigurationError("Unable to tie embeddings,"
+                                         "Selected source embedder is not an instance of `Embedding`.")
+            source_weight_shape = source_embedder.weight.shape
+            target_weight_shape = self._decoder.target_embedder.weight.shape
+            if  source_weight_shape != target_weight_shape:
+                raise ConfigurationError(f"Unable to tie embeddings due to mismatch in shared weights."
+                                         f"source  -  {source_weight_shape}"
+                                         f"target  -  {target_weight_shape}")
+            source_embedder.weight = self._decoder.target_embedder.weight
```

Comment:
Hmm, this seems potentially broken given that an `Embedding` also can have a `self._projection`. Is it possible to just share the exact same `TokenEmbedder`? Then we wouldn't have to make this specific to `Embedding` either.

Reply:
I can only set the decoder's embedder to the encoder's embedder  because the decoder embedder might have weights tied already with its output layer. Hence it cannot be overwritten.

Shall I  check if the `get_output_dim()` of both the embedders are same, and then just set the decoder embedder to the encoder key?

Reply:
Right, I didn't mean to reverse the direction that they're tied. My point was just that `self._projection` is currently being ignored which on the surface looks like a bug. Please clarify if I'm wrong!

> Shall I check if the `get_output_dim()` of both the embedders are same, and then just set the decoder embedder to the encoder key?

By that do you mean that you'd remove `tied_source_embedder_key`? I think we should keep it actually. There's no reason why the weights should always be tied just because the output dimension is the same.